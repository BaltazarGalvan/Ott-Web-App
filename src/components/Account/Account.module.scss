@use '../../styles/variables';
@use '../../styles/theme';
@use '../../styles/mixins/responsive';

button {
  margin-right: variables.$base-spacing;
}

.flexBox {
  display: flex;
  flex-direction: column;
  > * {
    margin-bottom: variables.$base-spacing / 2;
  }
}

.controls {
  margin-top: variables.$base-spacing;
}

.checkbox {
  position: relative;
  display: block;
  align-items: center;
  margin-top: variables.$base-spacing / 2;
  margin-bottom: 0;
  padding-left: variables.$base-spacing * 1.5;
  font-size: 16px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;

  > input {
    position: absolute;
    width: 0;
    height: 0;
    margin-right: 10px;
    opacity: 0;
    :not(&:disabled) {
      &:hover ~ .checkmark:hover {
        background-color: theme.$forms-checkbox-bg-hover;
        cursor: pointer;
      }
      &:hover ~ .checkLabel {
        cursor: pointer;
      }
      &:checked ~ .checkmark:hover {
        background-color: theme.$forms-primary-color-hover;
      }
    }
    &:checked ~ .checkmark {
      background-color: theme.$forms-primary-color;
    }
    &:checked ~ .checkmark::after {
      display: block;
    }

    &:disabled ~ .checkmark,
    &:disabled ~ .checkLabel {
      cursor: default;
      opacity: 0.6;
    }
  }
}

.checkmark {
  position: absolute;
  top: 0;
  left: 0;
  width: 15px;
  height: 15px;
  background-color: theme.$forms-checkbox-bg;
  &::after {
    content: '';
    position: absolute;
    top: 2px;
    left: 4px;
    display: none;

    width: 6px;
    height: 10px;
    border: solid black;
    border-width: 0 2px 2px 0;
    -webkit-transform: rotate(45deg);
    -ms-transform: rotate(45deg);
    transform: rotate(45deg);
  }
}

.submitConsents {
  margin-top: variables.$base-spacing;
}
