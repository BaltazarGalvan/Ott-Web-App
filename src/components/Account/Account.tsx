import React, { useState, useEffect, useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import type { Consent, Customer, CustomerConsent, UpdateCustomerPayload } from 'types/account';
import type { CustomerFormValues, FormErrors, GenericFormValues } from 'types/form';
import { useHistory } from 'react-router-dom';

import { formatConsentsFromValues, formatConsentValues } from '../../utils/collection';
import Visibility from '../../icons/Visibility';
import VisibilityOff from '../../icons/VisibilityOff';
import useToggle from '../../hooks/useToggle';
import Button from '../../components/Button/Button';
import Spinner from '../../components/Spinner/Spinner';
import Form from '../Form/Form';
import IconButton from '../IconButton/IconButton';
import LoadingOverlay from '../LoadingOverlay/LoadingOverlay';
import TextField from '../TextField/TextField';
import Checkbox from '../Checkbox/Checkbox';
import { addQueryParam } from '../../utils/history';

import styles from './Account.module.scss';

type Props = {
  customer: Customer;
  errors?: FormErrors<UpdateCustomerPayload>;
  isLoading: boolean;
  consentsLoading: boolean;
  publisherConsents?: Consent[];
  customerConsents?: CustomerConsent[];
  onUpdateEmailSubmit: (data: CustomerFormValues) => void;
  onUpdateInfoSubmit: (data: CustomerFormValues) => void;
  onDeleteAccountClick: () => void;
  onUpdateConsentsSubmit: (consents: CustomerConsent[]) => void;
  onReset?: () => void;
  panelClassName?: string;
  panelHeaderClassName?: string;
};

type Editing = 'none' | 'account' | 'password' | 'info' | 'consents';

const Account = ({
  customer,
  errors,
  isLoading,
  consentsLoading,
  publisherConsents,
  customerConsents,
  panelClassName,
  panelHeaderClassName,
  onUpdateEmailSubmit,
  onUpdateInfoSubmit,
  onDeleteAccountClick,
  onUpdateConsentsSubmit,
  onReset,
}: Props): JSX.Element => {
  const { t } = useTranslation('user');
  const history = useHistory();
  const [editing, setEditing] = useState<Editing>('none');
  const [viewPassword, toggleViewPassword] = useToggle();
  const consentValues = useMemo(() => formatConsentValues(publisherConsents, customerConsents), [publisherConsents, customerConsents]);
  const initialValues = useMemo(() => ({ ...customer, consents: consentValues }), [customer, consentValues]);

  const handleSubmit = (values: GenericFormValues) => {
    switch (editing) {
      case 'account':
        return onUpdateEmailSubmit(values as CustomerFormValues);
      case 'info':
        return onUpdateInfoSubmit(values as CustomerFormValues);
      case 'consents':
        return onUpdateConsentsSubmit(formatConsentsFromValues(publisherConsents, values));
      default:
        return;
    }
  };
  const onCancelClick = (formResetHandler?: () => void): void => {
    formResetHandler && formResetHandler();
    setEditing('none');
    onReset && onReset();
  };

  const editPasswordClickHandler = () => {
    history.push(addQueryParam(history, 'u', 'reset-password'));
  };

  useEffect(() => {
    !isLoading && setEditing('none');
  }, [isLoading]);

  return (
    <Form initialValues={initialValues} onSubmit={handleSubmit}>
      {({ values, handleChange, handleReset, handleSubmit, hasChanged }) => (
        <>
          {isLoading && <LoadingOverlay transparentBackground />}
          <div className={panelClassName}>
            <div className={panelHeaderClassName}>
              <h3>{t('account.email')}</h3>
            </div>
            <div className={styles.flexBox}>
              <TextField
                name="email"
                label={t('account.email')}
                value={values.email as string}
                onChange={handleChange}
                error={!!errors?.email}
                helperText={errors?.email}
                disabled={isLoading}
                editing={editing === 'account'}
              />
              {editing === 'account' && (
                <TextField
                  name="confirmationPassword"
                  label={t('account.confirm_password')}
                  value={values.confirmationPassword as string}
                  onChange={handleChange}
                  error={!!errors?.confirmationPassword}
                  helperText={errors?.confirmationPassword}
                  type={viewPassword ? 'text' : 'password'}
                  disabled={isLoading}
                  rightControl={
                    <IconButton
                      aria-label={viewPassword ? t('account.hide_password') : t('account.view_password')}
                      onClick={() => toggleViewPassword()}
                    >
                      {viewPassword ? <Visibility /> : <VisibilityOff />}
                    </IconButton>
                  }
                />
              )}
              <div className={styles.controls}>
                {editing === 'account' ? (
                  <>
                    <Button
                      label={t('account.save')}
                      type="submit"
                      onClick={handleSubmit}
                      disabled={isLoading || !values.email || !values.confirmationPassword}
                    />
                    <Button label={t('account.cancel')} type="reset" variant="text" onClick={() => onCancelClick(handleReset)} />
                    <Button label={t('account.delete_account')} type="button" onClick={onDeleteAccountClick} />
                  </>
                ) : (
                  <Button label={t('account.edit_account')} type="button" onClick={() => setEditing('account')} />
                )}
              </div>
            </div>
          </div>
          <div className={panelClassName}>
            <div className={panelHeaderClassName}>
              <h3>{t('account.security')}</h3>
            </div>
            <div>
              <strong>{t('account.password')}</strong>
              <p>****************</p>
              <Button label={t('account.edit_password')} type="button" onClick={() => (customer ? editPasswordClickHandler() : null)} />
            </div>
          </div>
          <div className={panelClassName}>
            <div className={panelHeaderClassName}>
              <h3>{t('account.about_you')}</h3>
            </div>
            <div>
              <div className={styles.flexBox}>
                <TextField
                  name="firstName"
                  label={t('account.firstname')}
                  value={values.firstName as string}
                  onChange={handleChange}
                  error={!!errors?.firstName}
                  helperText={errors?.firstName}
                  disabled={isLoading}
                  editing={editing === 'info'}
                />
                <TextField
                  name="lastName"
                  label={t('account.lastname')}
                  value={values.lastName as string}
                  onChange={handleChange}
                  error={!!errors?.lastName}
                  helperText={errors?.lastName}
                  disabled={isLoading}
                  editing={editing === 'info'}
                />
                <div className={styles.controls}>
                  {editing === 'info' ? (
                    <>
                      <Button label={t('account.save')} type="submit" disabled={!hasChanged} onClick={handleSubmit} />
                      <Button type="reset" variant="text" label={t('account.cancel')} onClick={() => setEditing('none')} />
                    </>
                  ) : (
                    <Button type="button" label={t('account.edit_information')} onClick={() => setEditing('info')} />
                  )}
                </div>
              </div>
            </div>
          </div>
          <div className={panelClassName}>
            <div className={panelHeaderClassName}>
              <h3>{t('account.terms_and_tracking')}</h3>
            </div>
            {consentsLoading ? (
              <Spinner size="small" />
            ) : publisherConsents ? (
              <div onClick={() => setEditing('consents')}>
                {publisherConsents.map((consent, index) => (
                  <Checkbox
                    key={index}
                    name={consent.name}
                    value={values.consents?.[consent.name] || ''}
                    checked={(values.consents?.[consent.name] as boolean) || false}
                    onChange={(event) => (handleChange ? handleChange(event, { nestInto: 'consents' }) : null)}
                    disabled={consent.required}
                    label={consent.label}
                  />
                ))}
                <Button
                  className={styles.submitConsents}
                  type="button"
                  label={t('account.update_consents')}
                  disabled={!hasChanged}
                  onClick={handleSubmit}
                />
              </div>
            ) : null}
          </div>
        </>
      )}
    </Form>
  );
};

export default Account;
