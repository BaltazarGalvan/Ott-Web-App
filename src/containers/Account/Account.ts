type ChildrenParams = {
  account: Account;
  update: () => void;
};

type Props = {
  children: (data: ChildrenParams) => JSX.Element;
};

const Account = ({ children }: Props): JSX.Element => {
  const account: Account = {};
  const update = (values: Account) => console.info('update', values);

  return children({ account, update } as ChildrenParams);
};

export default Account;
