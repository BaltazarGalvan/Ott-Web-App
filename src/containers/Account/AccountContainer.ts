import type { Consent, Customer, CustomerConsent, UpdateCustomerConsentsPayload, UpdateCustomerPayload } from 'types/account';
import { useMutation, useQuery } from 'react-query';
import type { CustomerFormErrors, CustomerFormValues, FormErrors } from 'types/form';

import { ConfigStore } from '../../stores/ConfigStore';
import { AccountStore } from '../../stores/AccountStore';
import { getCustomerConsents, getPublisherConsents, updateCustomer, updateCustomerConsents } from '../../services/account.service';

type ChildrenParams = {
  customer: Customer;
  errors: FormErrors<UpdateCustomerPayload>;
  isLoading: boolean;
  consentsLoading: boolean;
  publisherConsents?: Consent[];
  customerConsents?: CustomerConsent[];
  onUpdateEmailSubmit: (values: CustomerFormValues) => void;
  onUpdateInfoSubmit: (values: CustomerFormValues) => void;
  onUpdateConsentsSubmit: (consents: CustomerConsent[]) => void;
  onReset: () => void;
};

type Props = {
  children: (data: ChildrenParams) => JSX.Element;
  fetchConsents?: boolean;
};

const AccountContainer = ({ children, fetchConsents = true }: Props): JSX.Element => {
  const customer = AccountStore.useState((state) => state.user);
  const auth = AccountStore.useState((state) => state.auth);
  const { cleengId, cleengSandbox } = ConfigStore.useState((s) => s.config);
  const jwt = auth?.jwt || '';
  const publisherId = cleengId || '';
  const customerId = customer?.id || 0;

  const customerMutation = useMutation((values: CustomerFormValues) => updateCustomer(values, cleengSandbox, jwt));
  const { mutate: mutateCustomer, isLoading: isMutateCustomerLoading, data: mutateCustomerData, reset } = customerMutation;

  const mutation = useMutation((values: UpdateCustomerConsentsPayload) => updateCustomerConsents(values, cleengSandbox, jwt));
  const { mutate: mutateConsents, isLoading: isMutateConsentsLoading, data: mutateConsentsData } = mutation;

  const enabled = fetchConsents && !!publisherId && !!customer?.id;
  const fetchPublicherConsents = useQuery(['publisherConsents'], () => getPublisherConsents({ publisherId }, cleengSandbox), { enabled });
  const { data: publisherConsents, isLoading: publisherConsentsLoading } = fetchPublicherConsents;

  const fetchCustomerConsents = useQuery(['customerConsents'], () => getCustomerConsents({ customerId: customerId.toString() }, cleengSandbox, jwt), {
    enabled,
  });
  const { data: customerConsents, isLoading: customerConsentsLoading } = fetchCustomerConsents;

  const onUpdateEmailSubmit = ({ id, email, confirmationPassword }: CustomerFormValues) => mutateCustomer({ id, email, confirmationPassword });
  const onUpdateInfoSubmit = ({ id, firstName, lastName }: CustomerFormValues) => mutateCustomer({ id, firstName, lastName });
  const onUpdateConsentsSubmit = (consents: CustomerConsent[]) => mutateConsents({ id: customerId.toString(), consents });

  const translateErrors = (errors?: string[]) => {
    const formErrors: CustomerFormErrors = {};

    errors?.map((error) => {
      switch (error) {
        case 'Invalid param email':
          formErrors.email = 'Invalid email address!';
          break;
        case 'Customer email already exists':
          formErrors.email = 'Email already exists!';
          break;
        case 'Please enter a valid e-mail address.':
          formErrors.email = 'Please enter a valid e-mail address.';
          break;
        case 'Invalid confirmationPassword': {
          formErrors.confirmationPassword = 'Password incorrect!';
          break;
        }
        default:
          console.info('Unknown error', error);
          return;
      }
    });
    return formErrors;
  };

  return children({
    customer,
    isLoading: isMutateCustomerLoading || isMutateConsentsLoading,
    errors: translateErrors(mutateCustomerData?.errors || mutateConsentsData?.errors),
    publisherConsents: publisherConsents?.responseData?.consents,
    customerConsents: customerConsents?.responseData?.consents,
    consentsLoading: publisherConsentsLoading || customerConsentsLoading,
    onUpdateEmailSubmit,
    onUpdateInfoSubmit,
    onUpdateConsentsSubmit,
    onReset: reset,
  } as ChildrenParams);
};

export default AccountContainer;
