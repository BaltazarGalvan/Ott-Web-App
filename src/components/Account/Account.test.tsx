import React from 'react';
import { render } from '@testing-library/react';
import type { Customer } from 'types/account';

import Account from './Account';

describe('<Account>', () => {
  test('renders and matches snapshot', () => {
    const customer: Customer = {
      id: '1',
      email: 'todo@test.nl',
      locale: 'en_en',
      country: 'England',
      currency: 'Euro',
      lastUserIp: 'temp',
    };
    const { container } = render(<Account customer={customer} update={(customer) => console.info(customer)} />);

    // todo
    expect(container).toMatchSnapshot();
  });
});
