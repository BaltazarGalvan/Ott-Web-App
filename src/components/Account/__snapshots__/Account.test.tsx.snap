// Vitest Snapshot v1

exports[`<Account> > renders and matches snapshot 1`] = `
<div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.email
      </h3>
    </div>
    <div
      class="_flexBox_1c1c63"
    >
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_email"
        >
          account.email
        </label>
        <p>
          email@domain.com
        </p>
      </div>
    </div>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_account
        </span>
      </button>
    </div>
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.security
      </h3>
    </div>
    <div
      class="_flexBox_1c1c63"
    >
      <strong>
        account.password
      </strong>
      <p>
        ****************
      </p>
    </div>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_password
        </span>
      </button>
    </div>
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.about_you
      </h3>
    </div>
    <div
      class="_flexBox_1c1c63"
    >
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_firstName"
        >
          account.firstname
        </label>
        <p>
          John
        </p>
      </div>
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_lastName"
        >
          account.lastname
        </label>
        <p>
          Doe
        </p>
      </div>
    </div>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_information
        </span>
      </button>
    </div>
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.terms_and_tracking
      </h3>
    </div>
    <div
      class="_flexBox_1c1c63"
    />
    <div
      class="_controls_1c1c63"
    />
  </div>
  <div
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h3>
        account.other_registration_details
      </h3>
    </div>
    <div
      class="_flexBox_1c1c63"
    >
      <div
        class="_customFields_c968f7"
      />
    </div>
    <div
      class="_controls_1c1c63"
    />
  </div>
</div>
`;
